cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc

```diff
@@ -161,39 +173,88 @@ class PrimitiveFilterImpl {
       out_is_valid_ = out_arr->buffers[0]->mutable_data();
     }
     out_data_ = reinterpret_cast<T*>(out_arr->buffers[1]->mutable_data());
-    out_offset_ = out_arr->offset;
     out_length_ = out_arr->length;
+    DCHECK_EQ(out_arr->offset, 0);
     out_position_ = 0;
   }
 
-  void ExecNonNull() {
-    // Fast filter when values and filter are not null
-    ::arrow::internal::VisitSetBitRunsVoid(
-        filter_data_, filter_offset_, values_length_,
-        [&](int64_t position, int64_t length) { WriteValueSegment(position, length); });
+  void ExecREEFilter() {
+    if (filter_.child_data[1].null_count == 0 && values_null_count_ == 0) {
```

Comment:
I understand that, but this is copying the existing behavior (I just moved the code around a bit). 

And I think `GetNullCount` is avoided in the first place because forcing a scan of the bitmap to be able to use the slightly faster kernel in very few extra cases is not worth it.

Reply:
That might be the case indeed, though computing the null count should be extremely fast (it's a stream of popcount instructions).